python/lsst/skymap/packers.py

```diff
@@ -60,56 +144,163 @@ def getIntFromFilter(cls, name):
+            A config field whose instance values are [wrapper proxies to]
+            `SkyMapDimensionPackerConfig` instances.
+        """
+        return ConfigurableField(doc, target=cls.from_config, ConfigClass=cls.ConfigClass)
+
+    @classmethod
+    def from_config(
+        cls, data_id: DataCoordinate, config: SkyMapDimensionPackerConfig
+    ) -> SkyMapDimensionPacker:
+        """Construct a dimension packer from a config object and a data ID.
+
+        Parameters
+        ----------
+        data_id : `lsst.daf.butler.DataCoordinate`
+            Data ID that identifies at least the ``skymap`` dimension.  Must
+            have dimension records attached unless ``config.n_tracts`` and
+            ``config.n_patches`` are both not `None`.
+        config : `SkyMapDimensionPackerConfig`
+            Configuration object.
+
+        Returns
+        -------
+        packer : `SkyMapDimensionPackerConfig`
+            New dimension packer.
+
+        Notes
+        -----
+        This interface is provided for consistency with the `lsst.pex.config`
+        "Configurable" concept, and is invoked when ``apply(data_id)`` is
+        called on a config instance attribute that corresponds to a field
+        created by `make_config_field`.  The constructor signature cannot play
+        this role easily for backwards compatibility reasons.
+        """
+        return cls(
+            data_id.subset(data_id.universe.extract(["skymap"])),
+            n_bands=config.n_bands,
+            bands=config.bands,
+            n_tracts=config.n_tracts,
+            n_patches=config.n_patches,
+        )
 
     @property
     def maxBits(self) -> int:
         # Docstring inherited from DataIdPacker.maxBits
-        packedMax = self._tractPatchMax
-        if self._filterMax is not None:
-            packedMax *= self._filterMax
-        return packedMax.bit_length()
+        packedMax = self._n_tracts * self._n_patches
+        if self._n_bands:
+            packedMax *= self._n_bands
+        return (packedMax - 1).bit_length()
 
     def _pack(self, dataId: DataCoordinate) -> int:
         # Docstring inherited from DataIdPacker.pack
-        packed = dataId["patch"] + self._patchMax*dataId["tract"]
-        if self._filterMax is not None:
-            packed += self.getIntFromFilter(dataId["band"])*self._tractPatchMax
+        if dataId["patch"] >= self._n_patches:
+            raise ValueError(f"Patch ID {dataId['patch']} is out of bounds; expected <{self._n_patches}.")
+        if dataId["tract"] >= self._n_tracts:
+            raise ValueError(f"Tract ID {dataId['tract']} is out of bounds; expected <{self._n_tracts}.")
+        packed = dataId["patch"] + self._n_patches * dataId["tract"]
+        if self._n_bands:
+            if (band_index := self._bands.get(dataId["band"])) is None:
+                raise ValueError(
+                    f"Band {dataId['band']!r} is not supported by SkyMapDimensionPacker configuration."
```

Comment:
Please include the list of known bands here (`self._bands`). Having recently debugged things that depended on configs that I had to then hunt down, this is a convention we should have.